core/src/main/scala/zio/config/AnnotatedRead.scala

```diff
@@ -0,0 +1,23 @@
+package zio.config
+
+final case class AnnotatedRead[+A](value: A, annotations: Set[AnnotatedRead.Annotation]) { self =>
+  def map[B](f: A => B): AnnotatedRead[B] =
+    AnnotatedRead(f(value), annotations)
+
+  def mapError[E, B](f: A => Either[E, B]): Either[E, AnnotatedRead[B]] =
+    f(value) match {
+      case Left(value)  => Left(value)
+      case Right(value) => Right(AnnotatedRead(value, annotations))
+    }
+
+  def zip[B](that: AnnotatedRead[B]): AnnotatedRead[(A, B)] =
+    AnnotatedRead((value, that.value), self.annotations ++ that.annotations)
+}
+
+object AnnotatedRead {
+  sealed trait Annotation
+
+  object Annotation {
+    case object NonDefaultValue extends Annotation
```

Comment:
I'd change this to `DefaultValue` rather than the negative.

Reply:
@jdegoes 

Sure. Just that, we may have some additional logic in ReadModule to handle optional fields.

`NonDefaultValue`  corresponds to field that is either optional or has a default value, but user did provide some value for this field in the source .
Fo example, for the field of type `Option`, `NonDefaultValue` corresponds to `Some(x)` instead of None.

The annotation helps us identify products where user has provided some value for (some of the) optional fields, but forgot to provide the required fields. We fail at the optional node in the program if  `annotations.contains(NonDefaultValue) && numberOfRequiredFields == numberOfMissingValueError`

If we were storing only `DefaultValues`, then we need an additional computation of counting the number of optional fields and we fail if `(numberOfOptionalFields == annotations.filter(_ == NonDefaultValue)` and `numberOfRequiredFields == numberOfMissingValueError`.

Reply:
Which I think is fine too

Reply:
Yes, no worries, and maybe NonDefaultValue is the right choice. Maybe we can just call it `SuppliedValid` or something so it's not the negation of something.